src/utilities.ts

```diff
@@ -40,3 +41,39 @@ function sortAscending(left: Offer, right: Offer) {
 function sortDescending(left: Offer, right: Offer) {
   return sortAscending(left, right) * -1
 }
+
+interface ExecuteWithMetricsParams<T extends string> {
+  runnable: () => Promise<void>
+  totalCount: Counter<T>
+  errorsCount: Counter<T>
+  durationsTotals: Gauge<T>
+  labelValues?: LabelValues<T>
+  durations?: Gauge<T>
+  count?: Counter<T>
+}
+
+export async function executeWithMetrics<T extends string>(params: ExecuteWithMetricsParams<T>) {
```

Comment:
This practically begs to be turned into HOC closure wrapper
Like:
```ts
type MetricsParams<T extends string> = Omit<ExecuteWithMetricsParams<T>, 'runnable'>

export function withMetrics<T extends string>(params: MetricsParams<T>) {
  const { totalCount, count, errorsCount, durationsTotals, durations, labelValues } = params

  return <T extends (...args: any[]) => any>(runnable: T) =>
    async (...params: Parameters<T>): Promise<ReturnType<T>> => {
      // Total count, and count by labels
      totalCount.inc()
      if (count && labelValues) {
        count.inc(labelValues)
      }
      const endTimerTotal = durationsTotals.startTimer()
      const endTimer = durations?.startTimer()
      try {
        // Run
        return await runnable(...params)
      } catch (error) {
        // Count errors and rethrow
        errorsCount.inc()
        throw error
      } finally {
        // End timers
        endTimerTotal()
        if (endTimer && labelValues) {
          endTimer(labelValues)
        }
      }
    }
}


// ready-made HOCs
export const withGetOrderBookMetrics = withMetrics(getOrderBookMetrics)
export const withBuyAmountEstimationMetrics = withMetrics(buyAmountEstimationMetrics)
export const withOrderBookFetcherMetrics = withMetrics(orderBookFetcher)

// to be used like 
router.get('/markets/:base-:quote', withGetOrderBookMetrics(async (req, res) => {
  if (!req.query.atoms) {
    res.sendStatus(HTTP_STATUS_UNIMPLEMENTED)
    return
  }
  const serialized = orderbooksFetcher.serializeOrderbooks()
  const result = await pool.exec('markets', [
    serialized,
    req.params.base,
    req.params.quote,
    getHops(req, argv['max-hops']),
  ])
  res.json(result)
}))

// and in OrderbookFetcher::constructor
    const poll = withOrderBookFetcherMetrics(async () => {
      try {
        const contract = await this.loadBatchExchangeViewer()
        this.orderbooks = await updateOrderbooks(contract, pageSize, logger)
      } catch (error) {
        logger.error(`Failed to fetch Orderbooks: ${error}`, null)
      }
      this.intervalId = setTimeout(poll, pollFrequency)
    })
```

That way metrics logic would be isolated from actual express and orderbook logic.
If you decide to go this way, I can submit a PR to this PR <sub><sub><sub> PR-ception

Reply:
Yes, it's a good idea, do the honors @Velenir